### PR TITLE
Gracefully handle exceptions from Security Providers

### PR DESCRIPTION
**Pull request by jkakavas**
In certain situations, such as when configured in FIPS 140 mode,
the Java security provider in use might throw a subclass of
java.lang.Error. We currently do not catch these and as a result
the JVM exits, shutting down elasticsearch.

This commit attempts to address this by catching AssertionErrors
that we are aware that might be thrown when a PBKDF2 implementation
is used from a Security Provider in FIPS 140 mode, with the password
input being less than 14 bytes (112 bits).

- In our PBKDF2 family of hashers, we catch the AssertionError and
throw an ElasticsearchException while creating or verifying the
hash. We throw on verification instead of simply returning false
on purpose so that the message bubbles up and the cause becomes
obvious (otherwise it would be indistinguishable from a wrong
password).
- In KeyStoreWrapper, we catch the AssertionError in order to
wrap and rethrow with a helpful message. This can happen when using
any of the keystore CLI commands, when the node starts or when we attempt
to reload secure settings.
- In the `elasticsearch-users` tool, we catch the AssertionError and
throw an appropriate UserException.

Tests are missing because it's not trivial to set CI in fips approved mode
right now, and thus any tests would need to be muted. There is a parallel 
effort in https://github.com/elastic/elasticsearch/pull/64024 to enable that
and we can either hold off merging this until 64024 is merged and tests
added, or add tests in followup.

### CODE REVIEW

File: distribution/tools/keystore-cli/src/main/java/org/elasticsearch/common/settings/BaseKeyStoreCommand.java
Comment (jkakavas) [inline distribution/tools/keystore-cli/src/main/java/org/elasticsearch/common/settings/BaseKeyStoreCommand.java:71]: Because GeneralSecurityException can be thrown now too

File: server/src/main/java/org/elasticsearch/cli/KeyStoreAwareCommand.java
Comment (jkakavas) [inline server/src/main/java/org/elasticsearch/cli/KeyStoreAwareCommand.java:78]: Because GeneralSecurityException can be thrown now too

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java
Comment (jkakavas) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java]: My thinking is that we should scope this tightly first and expand if/when we get reports/CI failures for additional errors. The ones we know are thrown are all `AssertionError` and catching all `Error` felt too wide of a net. Not holding a strong opinion on this, happy to discuss

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java
Comment (jkakavas) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:544]: Throw instead of returning false so that this is distinguishable from wrong password, which i believe is desired.

File: server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java:317]: We throw `ElasticsearchException` in other places while throwing `GeneralSecurityException` here. Do we want to be consistent?
Nit: extra space (and the period?) at the end of the error message string :)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:544]: Nit: I cannot help trying to make this error message read more like the others. So how about "Error in using PBKDF2 implementation from the chosen Security Provider". Also this string has an extra period at the end while the above one does not :)

**Review by ywangd — COMMENTED** [review]
I left a few comments. The main question I have is whether we should be more liberal and catch `Throwable` instead of `AssertionError`, which is a choice of BC. Since we are trying to cover unexpected situations to avoid unpleasant surprises, I wonder whether widen the net is a viable option.

File: server/src/main/java/org/elasticsearch/cli/KeyStoreAwareCommand.java
Comment (jkakavas) [reply server/src/main/java/org/elasticsearch/cli/KeyStoreAwareCommand.java:78]: Yeah I missed that when I changed it in `BaseKeyStoreCommand`. readPassword throws a UserException and we don't want to catch it here either.

> Also KeyStoreWrapper#decrypt is declared to throw GeneralSecurityException as well. It wasn't a problem before or maybe it was just an oversight?

I think it was an oversight but it wasn't such an issue because the CLI would just exit throwing the exception. It wasn't as user friendly as a UserException with an error message but it wasn't too problematic to be noticed or reported, I'd guess

File: server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java
Comment (jkakavas) [reply server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java:316]: I went back and forth as to how wide the net should be. My counter argument is that when something throws an Error instead of an Exception it supposedly denotes an irrecoverable error. We _know_ that this is not true for the AssertionError that BCFIPS throws here, but can we be certain that we want to catch all `Throwable`s that can be thrown from external code. We don't do that in any other place. WDYT ?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java]: Somehow missed this comment in the previous pass ... I see you already considered it. I suggest a wider net (even wider then `Error`) because 1) any future fix to this would require a release which is purely overhead; 2) I cannot think of a valid reason for ES to shutdown here due to any kinda of error. So wider net (at least `Error`) should provide mostly benefit.

File: server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java
Comment (jkakavas) [reply server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java:317]: > We throw ElasticsearchException in other places while throwing GeneralSecurityException here. Do we want to be consistent?

We don't have consistency over this. I don't know if it should. Choice might seem arbitrary but I did put some thought in it.  I went with a GeneralSecurityException in the KeyStoreWrapper since it is an exception that is already used in similar issues here. I chose ElasticsearchException in Hasher as a generic elasticsearch related exception because I didn't find any one more specific that would make much sense. Happy to discuss and change that approach if you or Lyudmila have better suggestions

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java
Comment (jkakavas) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:544]: Sure, why not! I'll change the one below while we're at it

Comment (jkakavas) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:544]: I like `Error using PBKDF2 implementation from the selected Security Provider` slightly more if you don't mind :)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java
Comment (jkakavas) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java]: Thanks for the comment Yang. I see your points. I'll let Lyudmila weigh in too and we can reach a consensus

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java]: Sure. My comment on this was really my OCD.

**Comment by BigPandaToo** [comment]
Should we add a test?

**Comment by jkakavas** [comment]
> Should we add a test?

see description: 

> Tests are missing because it's not trivial to set CI in fips approved mode
right now, and thus any tests would need to be muted. There is a parallel
effort in #64024 to enable that
and we can either hold off merging this until 64024 is merged and tests
added, or add tests in followup.

File: server/src/main/java/org/elasticsearch/cli/KeyStoreAwareCommand.java
Comment (jkakavas) [reply server/src/main/java/org/elasticsearch/cli/KeyStoreAwareCommand.java:79]: This is addressed in https://github.com/elastic/elasticsearch/pull/65366 that is already merged

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java
Comment (BigPandaToo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java]: I would leave an ability to throw and shutdown for when something unthinkable happens. Catching `Error` should be ok though

**Comment by BigPandaToo** [comment]
LGTM

**Review by BigPandaToo — APPROVED** [review]

**Review by ywangd — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java
Comment (jkakavas) [reply server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java:316]: You got me thinking again @ywangd .. On the one hand, I tried to limit the scope of these try-catch blocks to a min so it makes sense that we would catch all irrecoverable errors from _that_ line and give out an error message. On the other hand, the very limited scope makes it less possible that a wider net would catch issues thrown by other security providers that we know nothing of (other than the PBKDF2 112 bits limitation). I'm a bit torn on this one

Comment (jkakavas) [reply server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java:316]: Had a short chat with Yang and we decided that catching `Error` is the way to go based on the arguments that have been made already in this PR

**Pull request merged**
